### ISSUE TITLE
Weird issue

### ISSUE THREAD

**Issue by mcseemk** [brunocodutra/ring-channel#35]
Consider this code:
```rust
pub async fn tst_nested(mut tx: ring_channel::RingSender<std::time::Instant>) {
    tx.send(std::time::Instant::now());
}

pub async fn tst(mut tx: ring_channel::RingSender<std::time::Instant>) {
    tx.send(std::time::Instant::now());
    tst_nested(tx.clone()).await;
}

fn main() {
    let mut rt = tokio::runtime::Runtime::new().unwrap();
    let (mut tx, mut rx) = ring_channel::ring_channel(std::num::NonZeroUsize::new(100).unwrap());
    rt.spawn(tst(tx));
    let msg = rx.recv().unwrap();
    let msg = rx.recv().unwrap();
}
```

It fails to compile with the following error:
```
error[E0277]: `std::ptr::NonNull<ring_channel::control::ControlBlock<std::time::Instant>>` cannot be shared between threads safely
   --> src/main.rs:147:8
    |
147 |     rt.spawn(tst(tx));
    |        ^^^^^ `std::ptr::NonNull<ring_channel::control::ControlBlock<std::time::Instant>>` cannot be shared between threads safely
    |
    = help: within `ring_channel::channel::RingSender<std::time::Instant>`, the trait `std::marker::Sync` is not implemented for `std::ptr::NonNull<ring_channel::control::ControlBlock<std::time::Instant>>`
    = note: required because it appears within the type `ring_channel::control::ControlBlockRef<std::time::Instant>`
```

I omitted the rest of the error description as it is long and scary.

The workarounds I found:
1. Do not call tst_nested() :)
2. Pass the RingSender by mutable reference. I'm not sure it is always possible/desireable though.
3. Just do not use tx.clone() when calling tst_nested() - for this tiny example it is OK, but generally not always possible.

I'm not 100% sure that it is a bug, but looks highly suspicious. Note that crossbeam_channel works fine in this scenario.

**Comment by mcseemk** [comment]
Another workaround: if channel is cloned before and then passed by value, it compiles fine.
So this code compiles:
```rust
    let tx2 = tx.clone();
    tst_nested(tx2).await;
```
While this one doesn't:
```rust
    tst_nested(tx.clone()).await;
```

Looks like a compiler bug now (?) I checked on beta and last nightly.

**Comment by brunocodutra** [comment]
Thanks for taking the time to report this.

#### TL;DR;

This is neither an issue with the compiler, nor with ring_channel. 

#### Why?

To understand why, consider the reduced snippet below, which reproduces the same error you see on `1.40.0-nightly (3a9abe3f0 2019-10-15)`.

```rust
async fn second(_: RingSender<u32>) {}

async fn first(tx: RingSender<u32>) {
    second(tx.clone()).await;
}

let mut pool = ThreadPool::new().unwrap();
let (tx, _) = ring_channel(std::num::NonZeroUsize::new(1).unwrap());
pool.spawn(first(tx)).unwrap();
```
Let us rewrite the line ` second(tx.clone()).await;` the way the compiler sees it:

```rust
async {
    let tx2 = tx.clone();
    second(tx2).await
}.await
```

Notice that `tx` is captured **by reference**, since its only use within the `async` block binds it to a shared reference in the call to `clone`. Recall that the `async` block is scheduled for execution in the `ThreadPool` and the corresponding task can be picked up by any thread, so `&tx` is required to implement `Send`, or equivalently, `tx` is required to implement `Sync`, [which it does not](https://docs.rs/ring-channel/0.7.0/ring_channel/struct.RingSender.html#impl-Sync), hence the compilation error.

#### Why does removing the call to `clone` avoid the error?

This is what the compiler sees in this case:

```rust
async {
    second(tx).await
}.await
```

Notice that `tx` is captured **by value**, since its only use  within the `async` block binds it to a `move` into `second`. Following the same reasoning around the fact the `async` block can be executed on any thread, we conclude that `tx` is only required to implement `Send`, [which it does](https://docs.rs/ring-channel/0.7.0/ring_channel/struct.RingSender.html#impl-Send), so the borrow checker is happy.

#### Why does cloning `tx` upfront work?

Once again, let's rewrite it the way the compiler sees it:

```rust
let tx2 = tx.clone();
async {
    second(tx2).await
}.await
```

Following the same logic, `tx2` is captured **by value**, which we have just verified to be fine.

#### Why does crossbeam_channel work when captured by reference?

Simply because [`crossbeam_channel::Sender` does implement `Sync`](https://docs.rs/crossbeam-channel/0.3.9/crossbeam_channel/struct.Sender.html#impl-Sync).

#### Why doesn't `ring_channel::Sender` implement Sync?

The way it's currently implemented does allow it to implement Sync, but I decided to reserve the right to change the implementation in the future in a way that breaks this property, considering it can just be cloned instead .
